### ISSUE TITLE
Should model output be class probabilities (should model end with softmax)?

### ISSUE THREAD

**Issue by vedal** [pytorch/captum#1060]
## ❓ Questions and Help

In most docstrings of captum attribution methods, there is an example where the model is said to return class probabilities: 
```
>>> # ImageClassifier takes a single input tensor of images Nx3x32x32,
>>> # and returns an Nx10 tensor of class probabilities.
```
However, the example net in captum tutorials such as [this](https://github.com/pytorch/captum/blob/master/tutorials/CIFAR_TorchVision_Interpret.ipynb) don't seem to have an `nn.Softmax` at the end. I also noted [this issue](https://github.com/pytorch/captum/issues/407#issue-639582944), where both with and without softmax is suggested. 

Finally, I noted that only `captum.AttributionVisualizer` [mentions](https://captum.ai/api/insights.html#attributionvisualizer) of a score function like softmax.

My question is: what is common practice: with or without softmax at the end of the network when computing attributions?

**Comment by aobo-y** [comment]
hi @vedal , it depends on your need. Captum, as a tool itself, is not opinionated. It is feasible to use either way.

Generally,  "with Softmax" is more reasonable for multi-class classification, if you want to explain the predicted probability. Because the probability is decided by the target logit under the context of all the other logits, the output should take all logits into account. However, this is eventually up to what you want to explain. You can still choose to attribute "without Softmax" if you want to understand how features impact that single logit, instead of the final probability.

Thank you pointing out the [CIFAR10 tutorial](https://github.com/pytorch/captum/blob/master/tutorials/CIFAR_TorchVision_Interpret.ipynb). I agree that having the "Softmax" there makes more sense (cc @NarineK).

**Comment by vedal** [comment]
@aobo-y thanks alot for the thorough answer! It makes sense!

Would it make sense, given that many models dont have softmax at the end during training since they use `nn.CrossEntropyLoss`, to have an argument `softmax=False` for each attribution-method, for optioanlly passing logits through a softmax? Especially since softmax seems to be suggested by IG author (see previous links)

**Comment by aobo-y** [comment]
@vedal what you said makes sense. But as Captum is trying to be a generic library for all kinds for models & problems, such an "append_softmax" option will likely only be useful for multi-class classification problem. It can be confusing to other users, say a regression model, and cause unnecessary misuse. We also would like to be less opinionated about what is the "common practice", especially the model architecture.

As you noticed, it is actually very easy for users to create a wrapper themselves, e.g.,
```py
def wrapper(*args, **kwargs):
  return torch.nn.functional.softmax(
    model(*args, **kwargs)
  )

ig = IntegratedGradient(wrapper)
```

I agree our doc & tutorials are not perfect. We can keep this issue open until I update them to be less confusing.

**Comment by vedal** [comment]
Thanks alot for your suggestion and nice answer @aobo-y. I appreciate it :)

**Comment by craymichael** [comment]
This issue appears to have been answered in the discussion above, so I will close it for now. Please feel free to reopen or follow up if anything is still unresolved.
